tools/run.py

```diff
@@ -0,0 +1,57 @@
+#!/usr/bin/env python3
+from string import Template
+from argparse import ArgumentParser
+from pathlib import Path
+import tempfile
+import sys
+import os
+
+
+def get_args():
+    parser = ArgumentParser()
+    parser.add_argument("-b", "--bpftrace", type=str, default="bpftrace",
+                        help="Path of the bpftrace executable")
+    parser.add_argument("-m", "--mmtk", type=str, required=True,
+                        help="Path of the MMTk binary")
+    parser.add_argument("-H", "--harness", action="store_true",
+                        help="Only collect data for the timing iteration (harness_begin/harness_end)")
+    parser.add_argument("-p", "--print-script", action="store_true",
+                        help="Print the content of the bpftrace script")
+    parser.add_argument(
+        "-f", "--format", choices=["text", "json"], default="text", help="bpftrace output format")
+    parser.add_argument("tool", type=str, help="Name of the bpftrace tool")
+    return parser.parse_args()
+
+
+def main():
+    args = get_args()
+    here = Path(__file__).parent.resolve()
+    bpftrace_script = here / f"{args.tool}.bt"
+    if not bpftrace_script.exists():
+        print(f"Tracing script {str(bpftrace_script)} not found.")
+        sys.exit(1)
+    mmtk_bin = Path(args.mmtk)
+    if not mmtk_bin.exists():
+        print(f"MMTk binary {str(mmtk_bin)} not found.")
+        sys.exit(1)
+    prologue_file = here / \
+        "prologue_with_harness.bt.fragment" if args.harness else "prologue_without_harness.bt.fragment"
+    with prologue_file.open() as fd:
+        prologue = fd.read()
+    with (here / "epilogue.bt.fragment").open() as fd:
+        epilogue = fd.read()
+    with bpftrace_script.open() as fd:
+        template = Template(prologue + fd.read() + epilogue)
+        with tempfile.NamedTemporaryFile(mode="w+t") as tmp:
+            content = template.safe_substitute(
+                MMTK=mmtk_bin, TMP_FILE=tmp.name)
+            if args.print_script:
+                print(content)
+            tmp.write(content)
+            tmp.flush()
+            os.execvp("sudo", ["sudo", args.bpftrace,
+                      "--unsafe", "-f", args.format, tmp.name])
```

Comment:
Python will automatically delete the `NamedTemporaryFile` when exiting the scope of `with`.  However, `execvp` never returns.  You can replace this with `subprocess.run` or `subprocess.Popen`.

An alternative to using temp file is using pipes.  I tried piping the script into `sudo bpftrace` and it works.  For example, I am currently using `erb` (from Ruby) for templating, and my command line looks like this:

```shell
erb record_every=50 so_path=/home/wks/projects/mmtk-github/openjdk/build/linux-x86_64-normal-server-release/jdk/lib/server/libmmtk_openjdk.so  gcvis.erb | sudo bpftrace - > openjdk-packet-size3-nopoll.log
```

In Python, [the `subprocess.run` function](https://docs.python.org/3/library/subprocess.html#subprocess.run) (or `popen.communicate`) allows you to specify the stdin as a pipe, so you can directly pipe the Python string `prologue + content + epilogue` into `sudo bpftrace`.

Reply:
That's a deliberate choice. I will add comments to explain this.